File: Project/javaScript/src/UserManager.js
/**
 * User operations: async authentication, registration, profile management.
 */

import { Mutex } from './Mutex.js';
import * as TaskRepo from './TaskRepository.js';
import {
  createUser,
  getProfile,
  updateProfile,
  validateUsername,
  validatePassword,
  hashPassword,
} from './User.js';

const mutex = new Mutex();

/**
 * @param {string} id
 * @returns {string}
 */
function generateId() {
  return `user_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
}

/**
 * Register a new user.
 * @param {string} username
 * @param {string} password
 * @param {{ displayName?: string, email?: string }} [profile]
 * @returns {Promise<{ success: boolean, user?: import('./User.js').UserProfile, error?: string }>}
 */
export async function register(username, password, profile = {}) {
  const u = validateUsername(username);
  if (!u.valid) return { success: false, error: u.error };
  const p = validatePassword(password);
  if (!p.valid) return { success: false, error: p.error };

  return mutex.runExclusive(async () => {
    const users = await TaskRepo.loadUsers();
    const normalized = username.trim().toLowerCase();
    if (users.some((x) => x.username.toLowerCase() === normalized)) {
      return { success: false, error: 'Username already taken' };
    }
    const passwordHash = await hashPassword(password);
    const user = createUser({
      id: generateId(),
      username: username.trim(),
      passwordHash,
      displayName: profile.displayName ?? username.trim(),
      email: profile.email ?? '',
    });
    users.push(user);
    await TaskRepo.saveUsers(users);
    return { success: true, user: getProfile(user) };
  });
}

/**
 * Authenticate user by username and password.
 * @param {string} username
 * @param {string} password
 * @returns {Promise<{ success: boolean, user?: import('./User.js').UserProfile, error?: string }>}
 */
export async function login(username, password) {
  const passwordHash = await hashPassword(password);
  return mutex.runExclusive(async () => {
    const users = await TaskRepo.loadUsers();
    const normalized = username.trim().toLowerCase();
    const user = users.find((u) => u.username.toLowerCase() === normalized);
    if (!user || user.passwordHash !== passwordHash) {
      return { success: false, error: 'Invalid username or password' };
    }
    return { success: true, user: getProfile(user) };
  });
}

/**
 * Get user by id (profile only).
 * @param {string} userId
 * @returns {Promise<import('./User.js').UserProfile | null>}
 */
export async function getById(userId) {
  return mutex.runExclusive(async () => {
    const users = await TaskRepo.loadUsers();
    const user = users.find((u) => u.id === userId);
    return user ? getProfile(user) : null;
  });
}

/**
 * Get user by username (internal; returns full user for assignee resolution).
 * @param {string} username
 * @returns {Promise<{ id: string, username: string } | null>}
 */
export async function getByUsername(username) {
  return mutex.runExclusive(async () => {
    const users = await TaskRepo.loadUsers();
    const normalized = username.trim().toLowerCase();
    const user = users.find((u) => u.username.toLowerCase() === normalized);
    return user ? { id: user.id, username: user.username } : null;
  });
}

/**
 * Update profile for a user (displayName, email).
 * @param {string} userId
 * @param {{ displayName?: string, email?: string }} updates
 * @returns {Promise<{ success: boolean, user?: import('./User.js').UserProfile, error?: string }>}
 */
export async function updateUserProfile(userId, updates) {
  return mutex.runExclusive(async () => {
    const users = await TaskRepo.loadUsers();
    const idx = users.findIndex((u) => u.id === userId);
    if (idx === -1) return { success: false, error: 'User not found' };
    users[idx] = updateProfile(users[idx], updates);
    await TaskRepo.saveUsers(users);
    return { success: true, user: getProfile(users[idx]) };
  });
}

export default {
  register,
  login,
  getById,
  getByUsername,
  updateUserProfile,
};

File: Project/java/src/main/java/com/basanta/todo/UserManager.java
package com.basanta.todo;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * User operations: register, login, profile (language-specific: Optional, Mutex).
 */
public final class UserManager {
    private static final Mutex mutex = new Mutex();

    private static String generateId() {
        return "user_" + System.currentTimeMillis() + "_" + Long.toHexString(Double.doubleToLongBits(Math.random()));
    }

    public static Result<User.UserProfile> register(String username, String password, String displayName, String email) {
        User.ValidationResult u = User.validateUsername(username);
        if (!u.valid) return Result.fail(u.error);
        User.ValidationResult p = User.validatePassword(password);
        if (!p.valid) return Result.fail(p.error);

        return mutex.runExclusive(() -> {
            try {
                List<User> users = new ArrayList<>(TaskRepository.loadUsers());
                String normalized = username.trim().toLowerCase();
                boolean exists = users.stream().anyMatch(x -> x.getUsername().toLowerCase().equals(normalized));
                if (exists) return Result.fail("Username already taken");
                String hash = User.hashPassword(password);
                User user = User.create(generateId(), username.trim(), hash,
                    displayName != null && !displayName.isBlank() ? displayName.trim() : username.trim(),
                    email != null ? email.trim() : "");
                users.add(user);
                TaskRepository.saveUsers(users);
                return Result.ok(user.toProfile());
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }

    public static Result<User.UserProfile> login(String username, String password) {
        String passwordHash = User.hashPassword(password);
        return mutex.runExclusive(() -> {
            try {
                List<User> users = TaskRepository.loadUsers();
                String normalized = username.trim().toLowerCase();
                User user = users.stream()
                    .filter(u -> u.getUsername().toLowerCase().equals(normalized))
                    .findFirst()
                    .orElse(null);
                if (user == null || !user.getPasswordHash().equals(passwordHash))
                    return Result.fail("Invalid username or password");
                return Result.ok(user.toProfile());
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }

    public static Optional<User.UserProfile> getById(String userId) {
        return mutex.runExclusive(() -> {
            try {
                return TaskRepository.loadUsers().stream()
                    .filter(u -> u.getId().equals(userId))
                    .findFirst()
                    .map(User::toProfile);
            } catch (IOException e) {
                return Optional.empty();
            }
        });
    }

    public static Result<User.UserProfile> updateUserProfile(String userId, String displayName, String email) {
        return mutex.runExclusive(() -> {
            try {
                List<User> users = new ArrayList<>(TaskRepository.loadUsers());
                int idx = -1;
                for (int i = 0; i < users.size(); i++) {
                    if (users.get(i).getId().equals(userId)) { idx = i; break; }
                }
                if (idx == -1) return Result.fail("User not found");
                User u = users.get(idx);
                users.set(idx, u.withProfile(displayName, email));
                TaskRepository.saveUsers(users);
                return Result.ok(users.get(idx).toProfile());
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }
}
